File: packages/runtime-ruby/test/lifecycle_test.rb
# frozen_string_literal: true
require 'minitest/autorun'
require 'stringio'
require 'copilotkit/intelligence'
require 'copilotkit/runtime'

class LifecycleTest < Minitest::Test
  def client(&handler)
    transport = Object.new
    transport.define_singleton_method(:request, &handler)
    CopilotKit::Intelligence.new(api_key: 'key', transport: transport)
  end

  def test_success_events_have_canonical_payloads_and_unsubscribe_is_idempotent
    seen = []
    sdk = client { |*_| { 'thread' => { 'id' => 'canonical' } } }
    listener = ->(thread) { seen << [:created, thread['id']] }
    unsubscribe = sdk.on_thread_created(&listener)
    sdk.on_thread_created(&listener)
    sdk.on_thread_updated { |thread| seen << [:updated, thread['id']] }
    sdk.on_thread_deleted { |event| seen << [:deleted, event] }

    sdk.create_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent')
    sdk.update_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent', updates: { name: 'New' })
    sdk.archive_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent')
    sdk.delete_thread(thread_id: 'thread/id+space', user_id: 'user', agent_id: 'agent')
    unsubscribe.call
    unsubscribe.call
    sdk.create_thread(thread_id: 'other', user_id: 'user', agent_id: 'agent')

    assert_equal [[:created, 'canonical'], [:updated, 'canonical'], [:updated, 'canonical'],
      [:deleted, { 'threadId' => 'thread/id+space', 'userId' => 'user', 'agentId' => 'agent' }]], seen
  end

  def test_listener_errors_preserve_success_and_other_listeners
    seen = []
    sdk = client { |*_| { 'thread' => { 'id' => 'canonical' } } }
    sdk.on_thread_created { |_| raise 'private application error' }
    sdk.on_thread_created { |thread| seen << thread }

    _, diagnostics = capture_io do
      assert_equal({ 'id' => 'canonical' }, sdk.create_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent'))
    end

    assert_equal [{ 'id' => 'canonical' }], seen
    assert_includes diagnostics, 'listener failed'
    refute_includes diagnostics, 'private application error'
  end

  def test_creation_events_distinguish_lookup_creation_and_conflict_recovery
    [[200], [404, 200], [404, 409, 200]].each do |statuses|
      expected = statuses == [404, 200] ? 1 : 0
      seen = []
      responses = statuses.dup
      sdk = client do |*_|
        status = responses.shift
        raise CopilotKit::Error.new(status, 'rejected') unless status == 200
        { 'thread' => { 'id' => 'canonical' } }
      end
      sdk.on_thread_created { |thread| seen << thread }

      sdk.get_or_create_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent')

      assert_equal expected, seen.length
    end
  end

  def test_denied_mutations_emit_no_success_events
    [403, 409, 503].each do |status|
      seen = []
      sdk = client { |*_| raise CopilotKit::Error.new(status, 'rejected') }
      sdk.on_thread_created { |event| seen << event }
      sdk.on_thread_updated { |event| seen << event }
      sdk.on_thread_deleted { |event| seen << event }

      assert_raises(CopilotKit::Error) { sdk.create_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent') }
      assert_raises(CopilotKit::Error) { sdk.update_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent', updates: {}) }
      assert_raises(CopilotKit::Error) { sdk.delete_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent') }

      assert_empty seen
    end
  end

  def test_non_thread_operations_and_invalid_responses_emit_no_success_events
    seen = []
    sdk = client { |*_| { 'thread' => { 'id' => 'thread' } } }
    sdk.on_thread_created { |event| seen << event }
    sdk.on_thread_updated { |event| seen << event }
    sdk.on_thread_deleted { |event| seen << event }
    [['POST', '/api/threads/subscribe'], ['PATCH', '/api/threads/thread/lock'],
     ['DELETE', '/api/threads/thread/lock'], ['GET', '/api/threads/thread'],
     ['POST', '/api/memories'], ['PATCH', '/api/memories/memory']].each do |method, path|
      sdk.request(method, path, { 'userId' => 'user', 'agentId' => 'agent' })
    end
    assert_empty seen
    [nil, [], {}, { 'thread' => {} }, { 'thread' => { 'id' => 3 } }].each do |response|
      malformed = client { |*_| response }
      malformed.on_thread_created { |event| seen << event }
      assert_raises(CopilotKit::Error) { malformed.create_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent') }
      assert_empty seen
    end
  end

  def test_block_is_required_and_listener_can_unsubscribe_itself
    sdk = client { |*_| { 'thread' => { 'id' => 'thread' } } }
    assert_raises(ArgumentError) { sdk.on_thread_created }
    seen = []
    unsubscribe = nil
    unsubscribe = sdk.on_thread_created { |event| seen << event; unsubscribe.call }

    2.times { sdk.create_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent') }

    assert_equal 1, seen.length
  end

  def test_rack_mutations_notify_the_shared_sdk_with_trusted_identity
    seen, payloads = [], []
    sdk = client { |_, _, body, _| payloads << body; { 'thread' => { 'id' => 'canonical' } } }
    sdk.on_thread_updated { |thread| seen << [:updated, thread['id']] }
    sdk.on_thread_deleted { |event| seen << [:deleted, event] }
    runtime = CopilotKit::Runtime.new(intelligence: sdk, identify_user: ->(_) { { 'id' => 'trusted' } },
      telemetry: CopilotKit::Telemetry.new(disabled: true))

    [['PATCH', '/threads/thread', { 'agentId' => 'agent', 'userId' => 'spoof', 'name' => 'New' }],
     ['POST', '/threads/thread/archive', { 'agentId' => 'agent' }],
     ['DELETE', '/threads/thread', { 'agentId' => 'agent' }]].each do |method, path, body|
      result = runtime.call('REQUEST_METHOD' => method, 'PATH_INFO' => path, 'QUERY_STRING' => '', 'rack.input' => StringIO.new(JSON.generate(body)))
      assert_equal 200, result.first
    end

    assert_equal [[:updated, 'canonical'], [:updated, 'canonical'], [:deleted, { 'threadId' => 'thread', 'userId' => 'trusted', 'agentId' => 'agent' }]], seen
    assert payloads.all? { |body| body['userId'] == 'trusted' }
  ensure
    runtime&.close
  end

  def test_concurrent_registration_deduplicates_the_same_listener
    sdk = client { |*_| { 'thread' => { 'id' => 'thread' } } }
    seen = Queue.new
    listener = ->(event) { seen << event }
    subscriptions = 8.times.map { Thread.new { sdk.on_thread_created(&listener) } }.map(&:value)

    sdk.create_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent')

    assert_equal 1, seen.size
    subscriptions.each(&:call)
    sdk.create_thread(thread_id: 'thread', user_id: 'user', agent_id: 'agent')
    assert_equal 1, seen.size
  end

  def test_run_notifies_persisted_creation_before_a_later_lock_failure
    seen, created = [], []
    sdk = client do |method, path, body, _|
      raise CopilotKit::Error.new(404, 'missing') if method == 'GET'
      if path == '/api/threads'
        created << body
        { 'thread' => { 'id' => 'canonical' } }
      else
        raise CopilotKit::Error.new(409, 'locked')
      end
    end
    sdk.on_thread_created { |thread| seen << thread }
    runtime = CopilotKit::Runtime.new(intelligence: sdk, identify_user: ->(_) { { 'id' => 'trusted' } },
      agents: { 'default' => ->(*) { flunk 'Agent must not run without a lock' } },
      learning_container: ->(*) { 'existing-container' }, telemetry: CopilotKit::Telemetry.new(disabled: true))

    result = runtime.call('REQUEST_METHOD' => 'POST', 'PATH_INFO' => '/agent/default/run', 'QUERY_STRING' => '',
      'rack.input' => StringIO.new(JSON.generate('threadId' => 'thread', 'runId' => 'run', 'messages' => [], 'tools' => [], 'context' => [], 'state' => {})))

    assert_equal 409, result.first
    assert_equal [{ 'id' => 'canonical' }], seen
    assert_equal [{ 'threadId' => 'thread', 'userId' => 'trusted', 'agentId' => 'default', 'learningContainerId' => 'existing-container' }], created
  ensure
    runtime&.close
  end
end

File: packages/runtime-go/intelligence/lifecycle_test.go
package intelligence_test

import (
	"context"
	"io"
	"net/http"
	"net/http/httptest"
	"reflect"
	"sync"
	"sync/atomic"
	"testing"

	"github.com/CopilotKit/CopilotKit/packages/runtime-go/intelligence"
)

func lifecycleClient(t *testing.T, handler http.HandlerFunc) *intelligence.Client {
	t.Helper()
	server := httptest.NewServer(handler)
	t.Cleanup(server.Close)
	client, err := intelligence.New(intelligence.Config{APIKey: "key", APIURL: server.URL})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(client.Close)
	return client
}

func TestLifecyclePayloadsAndUnsubscribe(t *testing.T) {
	client := lifecycleClient(t, func(w http.ResponseWriter, r *http.Request) {
		io.WriteString(w, `{"thread":{"id":"canonical","name":"Title"}}`)
	})
	var seen []string
	var deleted intelligence.ThreadDeletedPayload
	unsubscribe := client.OnThreadCreated(func(thread intelligence.Thread) { seen = append(seen, "created:"+thread.ID) })
	client.OnThreadUpdated(func(thread intelligence.Thread) { seen = append(seen, "updated:"+thread.ID) })
	client.OnThreadDeleted(func(event intelligence.ThreadDeletedPayload) { deleted = event; seen = append(seen, "deleted") })
	ctx := context.Background()

	if _, err := client.CreateThread(ctx, intelligence.CreateThreadParams{ThreadID: "thread", UserID: "user", AgentID: "agent"}); err != nil {
		t.Fatal(err)
	}
	if _, err := client.UpdateThread(ctx, "thread", "user", "agent", map[string]any{"name": "New"}); err != nil {
		t.Fatal(err)
	}
	if err := client.ArchiveThread(ctx, "thread", "user", "agent"); err != nil {
		t.Fatal(err)
	}
	if err := client.DeleteThread(ctx, "thread/id+space", "user", "agent"); err != nil {
		t.Fatal(err)
	}
	unsubscribe()
	unsubscribe()
	if _, err := client.CreateThread(ctx, intelligence.CreateThreadParams{ThreadID: "other", UserID: "user", AgentID: "agent"}); err != nil {
		t.Fatal(err)
	}

	if !reflect.DeepEqual(seen, []string{"created:canonical", "updated:canonical", "updated:canonical", "deleted"}) {
		t.Fatalf("unexpected events: %v", seen)
	}
	if deleted != (intelligence.ThreadDeletedPayload{ThreadID: "thread/id+space", UserID: "user", AgentID: "agent"}) {
		t.Fatalf("wrong deletion scope: %+v", deleted)
	}
}

func TestLifecyclePanicAndSelfUnsubscribe(t *testing.T) {
	client := lifecycleClient(t, func(w http.ResponseWriter, r *http.Request) { io.WriteString(w, `{"thread":{"id":"thread"}}`) })
	client.OnThreadCreated(func(intelligence.Thread) { panic("private application error") })
	count := 0
	var unsubscribe func()
	unsubscribe = client.OnThreadCreated(func(intelligence.Thread) { count++; unsubscribe() })

	for i := 0; i < 2; i++ {
		thread, err := client.CreateThread(context.Background(), intelligence.CreateThreadParams{ThreadID: "thread", UserID: "user", AgentID: "agent"})
		if err != nil || thread.ID != "thread" {
			t.Fatalf("listener replaced successful response: %v", err)
		}
	}

	if count != 1 {
		t.Fatalf("self unsubscribe or failure isolation failed: %d", count)
	}
}

func TestCreationNotificationsDistinguishConflicts(t *testing.T) {
	for _, statuses := range [][]int{{200}, {404, 200}, {404, 409, 200}} {
		calls := 0
		client := lifecycleClient(t, func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(statuses[calls])
			calls++
			io.WriteString(w, `{"thread":{"id":"thread"}}`)
		})
		count := 0
		client.OnThreadCreated(func(intelligence.Thread) { count++ })

		_, created, err := client.GetOrCreateThread(context.Background(), intelligence.CreateThreadParams{ThreadID: "thread", UserID: "user", AgentID: "agent"})

		if err != nil {
			t.Fatal(err)
		}
		expected := 0
		if len(statuses) == 2 {
			expected = 1
		}
		if count != expected || created != (expected == 1) {
			t.Fatalf("wrong notification for %v: %d %v", statuses, count, created)
		}
	}
}

func TestUnrelatedAndFailedWritesDoNotNotify(t *testing.T) {
	for _, status := range []int{200, 403, 409, 503} {
		client := lifecycleClient(t, func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(status)
			io.WriteString(w, `{"thread":{"id":"thread"}}`)
		})
		count := 0
		client.OnThreadCreated(func(intelligence.Thread) { count++ })
		client.OnThreadUpdated(func(intelligence.Thread) { count++ })
		client.OnThreadDeleted(func(intelligence.ThreadDeletedPayload) { count++ })
		operations := [][2]string{{"POST", "/api/threads/subscribe"}, {"PATCH", "/api/threads/thread/lock"}, {"DELETE", "/api/threads/thread/lock"}, {"GET", "/api/threads/thread"}, {"POST", "/api/memories"}}
		if status != 200 {
			operations = append(operations, [2]string{"POST", "/api/threads"}, [2]string{"PATCH", "/api/threads/thread"}, [2]string{"DELETE", "/api/threads/thread"})
		}

		for _, operation := range operations {
			_, err := client.Request(context.Background(), operation[0], operation[1], map[string]any{"userId": "user", "agentId": "agent"}, nil)
			if (err == nil) != (status == 200) {
				t.Fatalf("unexpected status outcome: %v", err)
			}
		}

		if count != 0 {
			t.Fatalf("unexpected event: %d", count)
		}
	}
}

func TestConcurrentLifecycleRegistration(t *testing.T) {
	client := lifecycleClient(t, func(w http.ResponseWriter, r *http.Request) { io.WriteString(w, `{"thread":{"id":"thread"}}`) })
	var count atomic.Int64
	var workers sync.WaitGroup
	for i := 0; i < 8; i++ {
		workers.Add(1)
		go func() {
			defer workers.Done()
			unsubscribe := client.OnThreadCreated(func(intelligence.Thread) { count.Add(1) })
			unsubscribe()
			unsubscribe()
		}()
	}
	workers.Wait()
	client.OnThreadCreated(func(intelligence.Thread) { count.Add(1) })

	for i := 0; i < 8; i++ {
		workers.Add(1)
		go func() {
			defer workers.Done()
			_, err := client.CreateThread(context.Background(), intelligence.CreateThreadParams{ThreadID: "thread", UserID: "user", AgentID: "agent"})
			if err != nil {
				t.Error(err)
			}
		}()
	}
	workers.Wait()

	if count.Load() != 8 {
		t.Fatalf("unexpected concurrent notifications: %d", count.Load())
	}
}

func TestMalformedThreadMutationDoesNotNotify(t *testing.T) {
	for _, body := range []string{"null", "[]", "{}", `{"thread":{}}`, `{"thread":{"id":3}}`} {
		client := lifecycleClient(t, func(w http.ResponseWriter, r *http.Request) { io.WriteString(w, body) })
		count := 0
		client.OnThreadCreated(func(intelligence.Thread) { count++ })

		_, err := client.CreateThread(context.Background(), intelligence.CreateThreadParams{ThreadID: "thread", UserID: "user", AgentID: "agent"})

		if err == nil || count != 0 {
			t.Fatalf("invalid response notified success: %q %v %d", body, err, count)
		}
	}
}
